payjoin/src/uri.rs

```diff
@@ -274,51 +359,154 @@ enum InternalPjParseError {
     UnsecureEndpoint,
 }
 
+/// Builder for `bip21::Uri` with `PayjoinParams`.
+///
+/// `amount` parameter is optional and can be changed with `[PjUriBuilder::amount]`.
+///
+/// `ohttp_config` parameter is required only for v2 payjoin.
+pub struct PjUriBuilder {
+    /// Address you want to receive funds to.
+    ///
+    /// Must be a valid bitcoin address.
+    address: Address,
+    /// Payjoing endpoint url listening for payjoin requests.
+    ///
+    /// Must be a valid url that can be parsed
+    /// with `[Payjoin::Url::parse]`.
+    pj_endpoint: Url,
+    /// Amount you want to receive.
+    ///
+    /// If `None` the amount will be left unspecified.
+    amount: Option<Amount>,
+    /// Message
+    message: Option<String>,
+    /// Label
+    pub(crate) label: Option<String>,
+    #[cfg(feature = "v2")]
+    /// Config for ohttp.
+    ///
+    /// `[PjUriBuilder::decode_ohttp_config]` can be used to convert
+    /// base64 string to `ohttp::KeyConfig`.
+    ///
+    /// Required only for v2 payjoin.
+    ohttp_config: ohttp::KeyConfig,
+}
+
+impl PjUriBuilder {
+    /// Create a new `PjUriBuilder` with required parameters.
+    pub fn new(
+        address: Address,
+        pj_endpoint: Url,
+        amount: Option<Amount>,
+        message: Option<String>,
+        label: Option<String>,
+        #[cfg(feature = "v2")] ohttp_config: ohttp::KeyConfig,
+    ) -> Self {
+        Self {
+            address,
+            pj_endpoint,
+            amount,
+            message,
+            label,
+            #[cfg(feature = "v2")]
+            ohttp_config,
+        }
+    }
+    pub fn build<'a>(self) -> PayjoinUri<'a, NetworkChecked, Payjoin> {
+        let pj_params = PayjoinParams {
+            endpoint: self.pj_endpoint,
+            disable_output_substitution: false,
+            #[cfg(feature = "v2")]
+            ohttp_config: Some(self.ohttp_config),
+        };
+        let pj_extras = Payjoin::Supported(pj_params);
+        let mut uri = PayjoinUri::with_extras(self.address, pj_extras);
+        let amount = self.amount.clone();
+        uri.set_amount(&amount);
+        uri.set_label(self.label.unwrap_or_default());
+        uri.set_message(self.message.unwrap_or_default());
+        uri
+    }
+}
+
+#[cfg(feature = "v2")]
+fn encode_ohttp_config(s: &ohttp::KeyConfig) -> Result<String, PjParseError> {
+    let res = s.encode().map_err(InternalPjParseError::BadOhttp)?;
+    let res = bitcoin::base64::encode_config(res, bitcoin::base64::URL_SAFE);
+    Ok(res)
+}
+
+// this might be needed when we want to implement Copy/Clone for ohttp::KeyConfig
+#[cfg(feature = "v2")]
+fn copy_ohttp_config(s: &ohttp::KeyConfig) -> Result<ohttp::KeyConfig, PjParseError> {
+    let res = s.encode().map_err(InternalPjParseError::BadOhttp)?;
```

Comment:
If a `payjoin::KeyConfig` could only be decoded from valid encoded type, then its `encode()` function should never fail. Such a type could `impl Clone` and solve the problem until[a `Clone` `ohttp::KeyConfig` is released 

tracking the release: https://github.com/martinthomson/ohttp/issues/55